TCES-67 Timeline Routing

kevinle623 opened the pull request:
# Important Note

Not sure what the plan is for the main menu dashboard so I left as is 

# Description

Implemented Timeline Feature for Employers, Clients, and Job Leads.

- able to view, filter + search and add for different timeline entries etc

Issue # TCES-67

## Gallary

![image](https://github.com/uoftblueprint/tces/assets/74465053/a6ac5261-2399-4e13-80f2-d4bf416a5b0c)

![image](https://github.com/uoftblueprint/tces/assets/74465053/5ee93687-178d-4b37-8bcd-8495a7fac78a)


![image](https://github.com/uoftblueprint/tces/assets/74465053/e263b68d-003b-499c-893f-abafebc4a7cf)


## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] 
# Checklist

Please delete options that are not applicable.

- [x] My code follows the style guidelines of this project (required)
- [x] I have performed a self-review of my code (required)
- [x] I have linted and formatted my code (required)
- [x] New and existing unit tests pass locally with my changes (required)
- [x] I have commented my code, particularly in hard-to-understand areas

annie951977 reviewed (changes requested):
The way you format your code makes it very readable, so that's awesome! 

I tried compiling your branch and aside from package lock stuff, I also got these errors. Let me know if you run into these problems as well or if there's something that's weird on my end

![image](https://user-images.githubusercontent.com/56055607/214842279-cc2163e6-fea7-4342-b17c-bee4a43b19f5.png)
![image](https://user-images.githubusercontent.com/56055607/214842318-ec379425-dc73-4729-9b32-263f749cbf80.png)
![image](https://user-images.githubusercontent.com/56055607/214842355-02580c37-2d45-4f54-b769-dc1a5b329704.png)

helen-li commented on client/src/App.tsx:
```diff
@@ -16,6 +18,8 @@ function App() {
 			<header className="App-header">
 				{Communication()}
+				{ActivityTimeline()}
+				{AddJobLead()}
```
Just for testing! I'm not entirely sure whether router is set up for client yet to navigate between pages so I just threw it in here, can totally remove before I merge

helen-li commented:
> The way you format your code makes it very readable, so that's awesome!
> 
> I tried compiling your branch and aside from package lock stuff, I also got these errors. Let me know if you run into these problems as well or if there's something that's weird on my end
> 
> ![image](https://user-images.githubusercontent.com/56055607/214842627-619c53b9-4f80-4714-b560-483cb7a5ff68.png) ![image](https://user-images.githubusercontent.com/56055607/214842318-ec379425-dc73-4729-9b32-263f749cbf80.png) ![image](https://user-images.githubusercontent.com/56055607/214842355-02580c37-2d45-4f54-b769-dc1a5b329704.png)

@Manal-jpg Could you try compiling on your end and see if you can reproduce these errors? I somehow could not reproduce the same warnings on my system and am concerned there's something wrong with my setup. Thanks!

For reference, my output just looks like this: 
<img src="https://user-images.githubusercontent.com/85378642/214896874-e01554f5-e6d1-4bac-af0b-b7aa956699d0.png">

Manal-jpg reviewed (approved):
I've finally reviewed the PR, sorry for the long wait. Everything looks great to me. I didn't get the errors Annie was having, and it was running on my computer without a problem (there was a small package error that Ganesh addressed on Slack). As for the employer conn.ts problem, I just realized that I accidentally removed my changes for that file while solving merge conflicts so sorry about that, and tysm for fixing it.

edwardhuahan commented:
I just compiled everything, it gave me a few warnings but no errors

![image](https://user-images.githubusercontent.com/17802344/217382196-f94b1997-9d2c-428a-9dff-639f25913ad5.png)

edwardhuahan commented:
Awesome job! It looks great, I think it'd be best to style your components using css files like we decided instead of the sx prop but just to save time I think it's fine this time

edwardhuahan reviewed (approved):
Make sure to make external css files next time for styling

annie951977 reviewed (approved):
Other that the previously addressed issues which look like are not replicable, looks good to me!

kevinle623 committed:
implemented job leads timeline

kevinle623 committed:
implemented client timeline

kevinle623 committed:
implemented employers timeline

kevinle623 committed:
fixed ui bugs

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-67-timeline-routing

kevinle623 committed:
routed job_lead_placement and added confirm dialog for case of overwriting placement

kevinle623 committed:
fixed lint

jordanjanakievski committed:
Fix frontend bugs for client card

jordanjanakievski committed:
Formatted files

jordanjanakievski commented:
Added some fixes to a frontend bug for the client page

DanielDervishi reviewed (commented):
Great work Kevin! Just a few things that I've noticed from clicking around:
- A timeline entry cuts into the search bar at times (Noticed this in the notes page of a client's timeline entry)
<img width="490" alt="Screenshot 2024-03-26 at 7 48 10 PM" src="https://github.com/uoftblueprint/tces/assets/58835213/9d2f8f77-6876-4a40-82a0-da2d0c15e2a3">

- A note as well as contact isn't showing up in the timeline after being added in the frontend in the client page.

- It seems updating a client's field adds a timeline entry for multiple fields on the client page.

- The date is a bit too small for the timeline entries

- Status at exit displays "no_results" rather than "No Results" (I assume this is the case for all status changes)

<img width="484" alt="Screenshot 2024-03-26 at 8 03 30 PM" src="https://github.com/uoftblueprint/tces/assets/58835213/721ba74d-fc47-4eec-b4f1-15edcc066c69">

- I think we should add the employer name alongside the joblead name just because it'll get a bit confusing when multiple companies have the same job lead available
- In the job lead timeline when placing a client, default to the existing job lead rather than showing a dropdown of job leads.

- Restrict placements such that there can't be more clients than positions available. So instead of kicking a client off of a job lead, just show a dialog saying all positions for this job lead have been filled or something along those lines.

- When switching from tab to tab, some additional timeline entries (or "No Placements Found") flash on the screen for like half a second. Is there any way to prevent this from happening?

- The search issue also exists on the job lead & employer timelines
<img width="543" alt="Screenshot 2024-03-26 at 8 23 27 PM" src="https://github.com/uoftblueprint/tces/assets/58835213/021a7361-0d9e-4b1b-a506-e13a9f7c8aaa">

- Also unrelated, when you click the back button on the job lead page there is a confirm exit dialog saying changes will not be save even when not editing

- Adding that you've contacted someone doesn't work on the employer timeline

- Lets restrict the jobs in the employer timeline just to those associated with the current employer


I'll take a closer look at the pr after these are addressed.

kevinle623 committed:
improved search functionality and only allow to add placements to job leads with enough head count

kevinle623 committed:
Merge branch 'kevinle6231/tces-67-timeline-routing' of https://github.com/uoftblueprint/tces into kevinle6231/tces-67-timeline-routing

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-67-timeline-routing

kevinle623 commented:
Hey Daniel thanks for the review here is progress / comments on the notes that you provided.

- A timeline entry cuts into the search bar at times (Noticed this in the notes page of a client's timeline entry) [FIXED]

- A note as well as contact isn't showing up in the timeline after being added in the frontend in the client page. [Can't seem to reproduce.]

- It seems updating a client's field adds a timeline entry for multiple fields on the client page. [FIXED]

- The date is a bit too small for the timeline entries [FIXED]

- Status at exit displays "no_results" rather than "No Results" (I assume this is the case for all status changes) [FIXED from now on, old entries will still have the underscore but from now on should be good]

- I think we should add the employer name alongside the joblead name just because it'll get a bit confusing when multiple companies have the same job lead available [Not sure what you mean by this]

- In the job lead timeline when placing a client, default to the existing job lead rather than showing a dropdown of job leads. [I currently have it so that it initially selects the current job lead, hope this is fine]

- Restrict placements such that there can't be more clients than positions available. So instead of kicking a client off of a job lead, just show a dialog saying all positions for this job lead have been filled or something along those lines. [I have it so that the job lead just doesn't show up in the dropdowns if it's full, hope this is fine, it was easy to implement too]

- When switching from tab to tab, some additional timeline entries (or "No Placements Found") flash on the screen for like half a second. Is there any way to prevent this from happening? [Every tab swtich is a new api call so I think it would be fine for now if thats fine.]

- The search issue also exists on the job lead & employer timelines [ FIXED]

- Also unrelated, when you click the back button on the job lead page there is a confirm exit dialog saying changes will not be save even when not editing [this is normal behaviour and easy to be implemented]

- Adding that you've contacted someone doesn't work on the employer timeline [Can't seem to reproduce the issue]

- Lets restrict the jobs in the employer timeline just to those associated with the current employer [This is the current behaviour]

kevinle623 committed:
fixed some bugs

kevinle623 committed:
fixed the client update

kevinle623 committed:
removed console.log

kevinle623 committed:
restrict the jobs in the employer timeline just to those associated with the current employer

kevinle623 committed:
fixed userID type error bug

DanielDervishi commented on client/src/components/client-page/client-timeline/index.jsx:
```diff
@@ -0,0 +1,140 @@
+      ...entryObject,
+      client: entryObject.object,
+    };
+    addEntry(clientEntryObject);
```
Add await since addEntry is async

DanielDervishi commented on client/src/components/client-page/client-timeline/client-timeline-entries/index.jsx:
```diff
@@ -0,0 +1,169 @@
+  };
+
+  useEffect(() => {
+    fetchClientTimelineEntries();
```
Be sure to pass globalSearchQuery here and to do the same thing for each of the other timelines

DanielDervishi commented on client/src/components/client-page/client-timeline/client-timeline-entries/index.jsx:
```diff
@@ -0,0 +1,169 @@
+  globalSearchQuery,
+}) {
+  const [timelineEntries, setTimelineEntries] = useState([]);
+  const [loading, setLoading] = useState(false);
```
Make setLoading true by default. This will help us avoid seeing "No timeline entries" flash on the screen when swapping tabs. Be sure to do this on all timelines.

DanielDervishi commented on client/src/App.jsx:
```diff
@@ -296,11 +300,22 @@ function App() {
+                    >
+                      <ClientsLoader setClients={setManagedClients}>
+                        <EmployersLoader setEmployers={setEmployers}>
+                          <EmployerPage
```
The current model for fetching objects in order to populate the timeline isn't really feasible. Tens of thousands of entries will be fetched each time one of these pages are loaded. If possible we should minimize data fetched by only loading objects used in the timeline when necessary or in smaller batches. I believe making use of backend search & pagination/loading results only when necessary  would be useful here. For example, I believe we can reduce job leads loaded to only those associated with the employer. Further, it might be best to wait until a user decides to initiate a placement to load the associated job leads/clients, serving them in batches upon the user's request while also allowing them to search by cleint/job lead name. The same applies to the loading done to accomodate the other timelines.

kevinle623 replied:
Hey @DanielDervishi, very good point thank for calling this out. I have fixed the fetching model such that we only fetch when the user decides to initiate such placement to load the assoicated job leads and clients. I made it so that for employer page timeline placement, you only get job leads that are asssoicated to that employer. In the Routing V2 ticket I will work towards a smarter pagination model in the dropdown combobox but for now I hope the current fix should be fine. :)

DanielDervishi reviewed (commented):
Hey Kevin, just left a few comments along with those sent on slack. We're really close here and should hopefully have this ticket merged in real soon :)

kevinle623 committed:
adjusted fetching model to be more on-demand fetching and other minor adjustments

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-67-timeline-routing

kevinle623 committed:
fixed bugs

kevinle623 committed:
fixed unit tests

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-67-timeline-routing

jordanjanakievski reviewed (commented):
For the most part, everything is looking good so far. I have not run into any issues with use.

For some of the add timeline entry options, the select is not auto-populating correctly. We can probably just automatically assign the value of the current object and not need to fetch all the clients/job leads
- Client Contact
- Job Placement (On Job Lead Page)
- Employer Contact
- Job Placement (On Employer Page -> Should just use the jobs linked with that employer. That data could be given outright to the page as it will be needed to create the card with all the employer's jobs)
#### I know these were discussed and working in some capacity earlier but may have broke with a later commit

Also after automatically assigning the field, we could disable it so it cannot be changed

kevinle623 committed:
resolved comments

kevinle623 commented:
> * Employer Contact

Hey Jordan,  thanks for feedback here is a recap of the suggested changes

- Client Contact [FIXED] - autoselected and fixed
- Job Placement (On Job Lead Page) [FIXED] - autoselected and fixed
- Employer Contact [FIXED] - autoselected and fixed
- Job Placement (On Employer Page -> Should just use the jobs linked with that employer. That data could be given outright to the page as it will be needed to create the card with all the employer's jobs) [FIXED] - job leads in the dropdown are fetched to only be linked to the employer

kevinle623 committed:
fixed tests

DanielDervishi reviewed (approved):
Amazing work Kevin! 
Huge props for sticking with this ticket and implementing all the feedback given.
I'm sure TCES will be really happy with the work you have done!

jordanjanakievski reviewed (approved):
Great job Kevin. This was a massive uptaking so thank you so much.
